### PR TITLE
Feature / Column Level Security

### PR DESCRIPTION
**Pull request by steve-keep**
## Summary by CodeRabbit

- **New Features**
	- Enhanced access control functionality with new methods for evaluating field-level permissions.
	- Introduction of a `RestrictedFieldError` for improved error handling related to GraphQL queries.
	- New automated tests for verifying column-level security and permissions in the application.
	- Added support for aggregate operations in data processing.

- **Bug Fixes**
	- Improved error logging messages for better clarity.

- **Documentation**
	- Updates to type definitions for better clarity and maintainability in access control mechanisms.

### CODE REVIEW

**Comment by steve-keep** [comment]
@coderabbitai review

File: src/packages/auth/src/authentication/apollo/plugin.ts [hunk 1]
```diff
@@ -264,6 +266,19 @@ export const authApolloPlugin = <R>(
...
+						// Here we are cleaning up the error messages to remove the empty data entity
+						errors = errors?.map((error: any) => {
+							if (error.extensions?.isRestrictedFieldError) {
+								delete error.path;
+								delete (response.body as any)?.singleResult.data;
+								delete error.extensions.isRestrictedFieldError;
+							}
+							return error;
+						});
+					}
```
[the remark this answers is not in the document]
Comment (steve-keep) [reply src/packages/auth/src/authentication/apollo/plugin.ts:281]: Unfortunately, in this instance I do need to delete the keys.

File: src/packages/auth/src/errors.ts [hunk 2]
```diff
@@ -23,3 +26,67 @@ export class ChallengeError extends ApolloError {
...
+	}
+
+	private formatFieldMessage() {
+		this.message = `Cannot query field "${this.field.name}" on type "${this.entityName}". [Suggestion hidden]?`;
+		this.extensions = {
+			...this.extensions,
+			code: RestrictedFieldErrorCode.GRAPHQL_VALIDATION_FAILED,
+			isRestrictedFieldError: true,
+		};
+	}
```
[the remark this answers is not in the document]
Comment (steve-keep) [reply src/packages/auth/src/errors.ts]: This is actually a really good comment 🎉

File: src/packages/auth/src/authentication/apollo/plugin.ts [hunk 3]
```diff
@@ -25,6 +25,8 @@ const didEncounterForbiddenError = (error: any): error is ForbiddenError =>
 	error.extensions.code === ErrorCodes.FORBIDDEN;
 const didEncounterChallengeError = (error: any): error is ChallengeError =>
 	error.extensions.code === ErrorCodes.CHALLENGE;
+const didEncounterRestrictedFieldError = (error: any): error is RestrictedFieldError =>
+	error.extensions?.isRestrictedFieldError;
```
Comment (thekevinbrown) [inline src/packages/auth/src/authentication/apollo/plugin.ts:29]: Should we put the `?` on the other ones above this for good measure while we're in here?

File: src/packages/auth/src/decorators/hooks/acl.ts [hunk 4]
```diff
@@ -119,22 +132,44 @@ const assertUserCanPerformRequest = async <G>(
...
 	entityMetadata: EntityMetadata<G>,
 	requestedFields: ResolveTree
 ) => {
-	const permissionsList: RequiredPermission[] = [`${entityMetadata.name}:${AccessType.Read}`];
+	const permissionsList: RequiredPermission[] = [
+		{
+			entityName: entityMetadata.name,
+			accessType: AccessType.Read,
+			type: RequirePermissionType.ENTITY,
+		},
```
Comment (thekevinbrown) [inline src/packages/auth/src/decorators/hooks/acl.ts:162]: 👨‍🍳 💋

File: src/packages/auth/src/errors.ts [hunk 5]
```diff
@@ -23,3 +26,63 @@ export class ChallengeError extends ApolloError {
...
+				break;
+			case FieldLocation.NESTED_FILTER:
+				this.formatFilterMessage();
+				break;
+			case FieldLocation.FILTER:
+				this.formatArgMessage();
+				break;
+			case FieldLocation.INPUT:
+				this.formatArgMessage();
+				break;
```
Comment (thekevinbrown) [inline src/packages/auth/src/errors.ts:61]: Do we need a `default:` here, or is the default already good enough?

**Review by thekevinbrown — APPROVED** [review]

File: src/packages/auth/src/errors.ts (hunk 5 above)
Comment (steve-keep) [reply src/packages/auth/src/errors.ts:61]: 🤔 Should we error maybe if it does not match?

**Pull request merged**